crates/matrix-sdk/src/event_cache/redecryptor.rs

```diff
@@ -0,0 +1,208 @@
+    Client,
+    event_cache::{
+        EventCache, EventCacheError, EventCacheInner, EventsOrigin, RoomEventCacheUpdate,
+    },
+};
+
+impl EventCache {
+    async fn get_utds(
+        &self,
+        room_key_info: &RoomKeyInfo,
+    ) -> Result<Vec<(OwnedEventId, Raw<AnySyncTimelineEvent>)>, EventCacheError> {
+        let map_timeline_event = |event: TimelineEvent| {
+            let event_id = event.event_id();
+
+            // Only pick out events that are UTDs, get just the Raw event as this is what
+            // the OlmMachine needs.
+            let event =
+                as_variant!(event.kind, TimelineEventKind::UnableToDecrypt { event, .. } => event);
+            // Zip the event ID and event together so we don't have to pick out the event ID
+            // again. We need the event ID to replace the event in the cache.
+            event_id.zip(event)
+        };
+
+        // Load the relevant events from the event cache store and attempt to redecrypt
+        // things.
+        let store = self.inner.store.lock().await?;
+        let events = store
+            .get_room_events(
+                &room_key_info.room_id,
+                Some("m.room.encrypted"),
+                Some(&room_key_info.session_id),
+            )
+            .await?;
+
+        Ok(events.into_iter().filter_map(map_timeline_event).collect())
+    }
+
+    #[instrument(skip_all, fields(room_id))]
+    async fn on_resolved_utds(
+        &self,
+        room_id: &RoomId,
+        events: Vec<(OwnedEventId, DecryptedRoomEvent)>,
+    ) -> Result<(), EventCacheError> {
+        // Get the cache for this particular room and lock the state for the duration of
+        // the decryption.
+        let (room_cache, _drop_handles) = self.for_room(room_id).await?;
+        let mut state = room_cache.inner.state.write().await;
+
+        let event_ids: Vec<_> = events.iter().map(|(event_id, _)| event_id).collect();
+
+        trace!(?event_ids, "Replacing successfully re-decrypted events");
+
+        for (event_id, decrypted) in events {
+            // The event isn't in the cache, nothing to replace. Realistically this can't
+            // happen since we retrieved the list of events from the cache itself and
+            // `find_event()` will look into the store as well.
+            if let Some((location, mut target_event)) = state.find_event(&event_id).await? {
+                target_event.kind = TimelineEventKind::Decrypted(decrypted);
+
+                // TODO: `replace_event_at()` propagates changes to the store for every event,
+                // we should probably have a bulk version of this?
+                state.replace_event_at(location, target_event).await?
+            }
+        }
+
+        // We replaced a bunch of events, reactive updates for those replacements have
+        // been queued up. We need to send them out to our subscribers now.
+        let diffs = state.room_linked_chunk_mut().updates_as_vector_diffs();
```

Comment:
Note that this is safe only because we have a write-lock over the `RoomEventCache` at the beginning. Otherwise, the risk would have been to read updates from another async operation. All good then.